### PR TITLE
Don't reserve new blocks if capacity is sufficient

### PR DESCRIPTION
**Pull request by thomwiggers**
By checking the current length we'll avoid doing many mallocs.

### CODE REVIEW

**Comment by thomwiggers** [comment]
Without this, this snippet of code will quickly use up all available RAM:

```
let v = Vob::with_capacity(10);
for _ in 0..100 {
    v.clear();
    v.extend_from_vob(&Vob::from_elem(3, false));
}

File: src/lib.rs [hunk 1]
```diff
@@ -247,16 +247,18 @@ impl<T: Debug + PrimInt + One + Zero> Vob<T> {
...
-        // used, not how many bits of capacity the user thinks we have. For example, repeatedly
-        // asking for storage for one extra bit will repeatedly add a whole extra machine word.
-        // Fixing this would require an extra machine word in the Vob struct, which hardly seems
-        // worth it for an unlikely, and probably silly, use case.
-        let cap = self.vec.capacity()
-                          .checked_mul(bits_per_block::<T>())
-                          .and_then(|x| x.checked_add(additional))
-                          .expect("Overflow detected");
-        self.vec.reserve(blocks_required::<T>(cap));
+        let current_free_space: usize = self.capacity() - self.len();
```
Comment (ltratt) [inline src/lib.rs]: Should we call this something shorter like `unused_bits`? Or even just `ub` as in one of the other functions.

File: src/lib.rs [hunk 2]
```diff
@@ -247,16 +247,18 @@ impl<T: Debug + PrimInt + One + Zero> Vob<T> {
...
+        if current_free_space > additional {
```
Comment (ltratt) [inline src/lib.rs]: We can simplify this to:

```rust
if additional > ub {
    // what is currently the else branch
}
```

**Comment by ltratt** [comment]
A couple of minor comments, but we should be able to get this in quickly. I'm a bit disappointed I didn't think of doing this originally (instead of writing a long comment explaining why it wasn't worth doing...)!

File: src/lib.rs (hunk 1 above)
Comment (thomwiggers) [reply src/lib.rs]: I've done `unused_bits`, as `ub` turned out to mean `used_bits` in most places I checked. It's sometimes a bit confusing as doing all this bit-mangling gives me headaches.

**Comment by thomwiggers** [comment]
This one also already got fixed up, by the way.

**Comment by ltratt** [comment]
If you squash this as you feel appropriate, I can merge it.

**Comment by thomwiggers** [comment]
Rebased.

**Comment by ltratt** [comment]
Thanks!

**Pull request merged**